File: events.py
"""
이벤트 버스: 분석기 → (우선순위 큐) → 보이스 워커.

분석기는 Event를 push만 한다. 큐가 다음을 책임진다:
  - 우선순위 정렬 (CRITICAL이 항상 먼저 나감)
  - 유형별 쿨다운 (같은 얘기 반복 방지 — 발화 억제 규칙의 일부)
  - 중복 억제 (같은 dedup_key가 대기 중이면 무시)
  - TTL (오래돼서 의미 없어진 이벤트 폐기 — 트래픽 콜 등)
"""

from __future__ import annotations

import heapq
import itertools
import logging
import threading
import time
from dataclasses import dataclass, field
from enum import IntEnum
from typing import Callable, Optional

log = logging.getLogger("events")


class Priority(IntEnum):
    CRITICAL = 0    # 즉시 발화, 재생 중인 저우선순위 멘트 중단 가능
    HIGH = 1        # 다음 차례에 발화
    NORMAL = 2      # 여유 있을 때 발화 (LLM 멘트 등)


# 이벤트 타입 상수 (쿨다운 키로도 사용)
class EventType:
    FUEL_CRITICAL = "fuel_critical"
    FUEL_WARNING = "fuel_warning"
    PIT_CALL = "pit_call"
    TRAFFIC_APPROACH = "traffic_approach"
    TRAFFIC_CLOSE = "traffic_close"
    TRAFFIC_UPDATE = "traffic_update"       # 상태 전이 후속 (지나감/떨어짐)
    TRAFFIC_MULTI = "traffic_multi"         # 다중 차량 종합 한 문장
    BRIDGE_FOLLOWUP = "bridge_followup"     # 긴급 콜 뒤 LLM 후속 설명
    DAMAGE = "damage"
    PENALTY = "penalty"
    PACE_COMMENT = "pace_comment"
    GAP_COMMENT = "gap_comment"
    TYRE_WARNING = "tyre_warning"
    LAP_ANALYSIS = "lap_analysis"       # LLM 판단형 멘트 (v0.4)
    STINT_BRIEFING = "stint_briefing"   # LLM (v0.4)
    SESSION_BRIEFING = "session_briefing"  # 세션 시작 브리핑 (세션당 1회)
    RACE_START = "race_start"
    RACE_END = "race_end"
    LAP_FEEDBACK = "lap_feedback"       # 트레이닝: 섹터 델타 피드백
    TRACK_TREND = "track_trend"         # 트레이닝: 과거 세션 대비 추세
    FCY = "fcy"                         # 풀코스옐로/세이프티카 시작
    FCY_PIT_OPEN = "fcy_pit_open"       # FCY 중 피트 오픈 (전략 기회)
    GREEN_FLAG = "green_flag"           # 리스타트/그린
    SECTOR_YELLOW = "sector_yellow"     # 로컬 옐로
    BLUE_FLAG = "blue_flag"             # 블루 플래그 (랩 앞선 차에 양보)
    RACE_MILESTONE = "race_milestone"   # 남은 시간/랩 카운트다운, 마지막 랩
    POSITION_CHANGE = "position_change" # 클래스 순위 변동
    RIVAL_PIT = "rival_pit"             # 동클래스 경쟁자 피트 (언더컷/오버컷)
    RIVAL_PACE = "rival_pace"           # 라이벌 페이스 비교 인텔
    PIT_LIMITER = "pit_limiter"         # 피트레인 리미터 미작동 경고
    ENGINE_WARNING = "engine_warning"   # 수온/유온/과열
    BRAKE_WARNING = "brake_warning"     # 브레이크 온도
    FUEL_SAVE = "fuel_save"             # 연료 세이브 목표 코칭


# 타입 → 쿨다운 설정 키 매핑 (config.cooldowns)
COOLDOWN_KEY = {
    EventType.FUEL_CRITICAL: "fuel_warning",
    EventType.FUEL_WARNING: "fuel_warning",
    EventType.TRAFFIC_APPROACH: "traffic",
    EventType.TRAFFIC_CLOSE: "traffic_close",
    EventType.TRAFFIC_UPDATE: "traffic_update",
    EventType.TRAFFIC_MULTI: "traffic_multi",
    EventType.BRIDGE_FOLLOWUP: "bridge",
    EventType.PACE_COMMENT: "pace_comment",
    EventType.GAP_COMMENT: "gap_comment",
    EventType.TYRE_WARNING: "tyre_warning",
    EventType.DAMAGE: "damage",
    EventType.PENALTY: "penalty",
    EventType.FCY: "race_control",
    EventType.FCY_PIT_OPEN: "fcy_pit_open",   # 전용 버킷 — FCY 콜에 눌리면 안 됨
    EventType.GREEN_FLAG: "green_flag",
    EventType.SECTOR_YELLOW: "sector_yellow",
    EventType.BLUE_FLAG: "blue_flag",
    EventType.RACE_MILESTONE: "race_milestone",
    EventType.POSITION_CHANGE: "position_change",
    EventType.RIVAL_PIT: "rival_pit",
    EventType.RIVAL_PACE: "rival_pace",
    EventType.PIT_LIMITER: "pit_limiter",
    EventType.ENGINE_WARNING: "engine_warning",
    EventType.BRAKE_WARNING: "engine_warning",
    EventType.FUEL_SAVE: "fuel_save",
}

# CRITICAL 쿨다운 예외: 더 심각한 단계는 쿨다운 무시하고 1회 통과
DEFAULT_TTL = {
    Priority.CRITICAL: 8.0,     # 긴급 콜은 8초 지나면 의미 없음
    Priority.HIGH: 20.0,
    Priority.NORMAL: 45.0,
}


@dataclass
class Event:
    type: str
    priority: Priority
    data: dict = field(default_factory=dict)   # 멘트 생성에 쓸 가공된 값
    message: Optional[str] = None              # 이미 완성된 멘트 텍스트 (있으면 그대로 사용)
    dedup_key: Optional[str] = None            # 없으면 type이 dedup 키
    ttl: Optional[float] = None
    tone: str = "casual"                       # casual | urgent — 변형 풀 톤 선택
    bridge: Optional[dict] = None              # 긴급 콜 뒤 LLM 후속 생성 컨텍스트
    valid_fn: Optional[Callable[[], bool]] = None  # 재생 직전 유효성 검사 (False→폐기)
    created_at: float = field(default_factory=time.monotonic)

    @property
    def key(self) -> str:
        return self.dedup_key or self.type

    def expired(self, now: float) -> bool:
        ttl = self.ttl if self.ttl is not None else DEFAULT_TTL[self.priority]
        return now - self.created_at > ttl


class EventBus:
    def __init__(self, cooldowns: dict):
        self._cooldowns = cooldowns
        self._heap: list = []
        self._counter = itertools.count()
        self._pending_keys: set[str] = set()
        self._last_fired: dict[str, float] = {}   # 쿨다운키 → 마지막 수락 시각
        self._lock = threading.Lock()
        self._available = threading.Event()
        self.urgent_pending = threading.Event()   # 재생 중단 판단용

    def _cooldown_sec(self, etype: str) -> float:
        key = COOLDOWN_KEY.get(etype, etype)
        return float(self._cooldowns.get(key, self._cooldowns.get("default", 60)))

    def push(self, ev: Event) -> bool:
        """수락되면 True. 쿨다운/중복으로 버려지면 False."""
        now = time.monotonic()
        with self._lock:
            if ev.key in self._pending_keys:
                return False
            cd_key = COOLDOWN_KEY.get(ev.type, ev.type)
            last = self._last_fired.get(cd_key)
            if last is not None and now - last < self._cooldown_sec(ev.type):
                # CRITICAL은 같은 쿨다운 그룹의 하위 우선순위보다 한 단계 봐준다:
                # 쿨다운 절반이 지났으면 통과 (연료 warning 직후 critical 등)
                if not (ev.priority == Priority.CRITICAL
                        and now - last >= self._cooldown_sec(ev.type) * 0.5):
                    log.debug("쿨다운으로 무시: %s", ev.type)
                    return False
            self._last_fired[cd_key] = now
            self._pending_keys.add(ev.key)
            heapq.heappush(self._heap, (int(ev.priority), next(self._counter), ev))
            self._available.set()
            if ev.priority == Priority.CRITICAL:
                self.urgent_pending.set()
        log.debug("이벤트 수락: %s (%s)", ev.type, ev.priority.name)
        return True

    def pop(self, timeout: float = 0.5) -> Optional[Event]:
        """우선순위가 가장 높은 유효 이벤트를 꺼낸다. 없으면 None."""
        if not self._available.wait(timeout):
            return None
        now = time.monotonic()
        with self._lock:
            while self._heap:
                _, _, ev = heapq.heappop(self._heap)
                self._pending_keys.discard(ev.key)
                if ev.expired(now):
                    log.debug("TTL 만료로 폐기: %s", ev.type)
                    continue
                if ev.valid_fn is not None:
                    try:
                        if not ev.valid_fn():
                            log.debug("상황 종료로 폐기: %s", ev.type)
                            continue
                    except Exception:
                        pass  # 유효성 검사 실패는 재생 강행보다 안전한 쪽(폐기 안 함)
                if not self._heap:
                    self._available.clear()
                if not any(e.priority == Priority.CRITICAL for _, _, e in self._heap):
                    self.urgent_pending.clear()
                return ev
            self._available.clear()
            self.urgent_pending.clear()
        return None

    def clear(self) -> None:
        """세션 경계에서 호출 — 대기 이벤트와 쿨다운 기록을 모두 비운다."""
        with self._lock:
            self._heap.clear()
            self._pending_keys.clear()
            self._last_fired.clear()
            self._available.clear()
            self.urgent_pending.clear()

File: simhub/src/TeamRadio56.Core/Logic/Events.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Threading;

namespace TeamRadio56.Core.Logic
{
    /// <summary>events.py 포팅 — 우선순위/이벤트 타입/이벤트 버스.</summary>
    public enum Priority
    {
        Critical = 0,   // 즉시 발화, 재생 중인 저우선순위 멘트 중단 가능
        High = 1,       // 다음 차례에 발화
        Normal = 2,     // 여유 있을 때 발화 (LLM 멘트 등)
    }

    /// <summary>이벤트 타입 상수 (쿨다운 키로도 사용). events.py의 EventType.</summary>
    public static class EventTypes
    {
        public const string FuelCritical = "fuel_critical";
        public const string FuelWarning = "fuel_warning";
        public const string PitCall = "pit_call";
        public const string TrafficApproach = "traffic_approach";
        public const string TrafficClose = "traffic_close";
        public const string TrafficUpdate = "traffic_update";
        public const string TrafficMulti = "traffic_multi";
        public const string Spotter = "spotter";
        public const string BridgeFollowup = "bridge_followup";
        public const string Damage = "damage";
        public const string DamageReport = "damage_report";
        public const string WheelDamage = "wheel_damage";
        public const string PartDetached = "part_detached";
        public const string Penalty = "penalty";
        public const string PaceComment = "pace_comment";
        public const string GapComment = "gap_comment";
        public const string TyreWarning = "tyre_warning";
        public const string LapAnalysis = "lap_analysis";
        public const string StintBriefing = "stint_briefing";
        public const string SessionBriefing = "session_briefing";
        public const string LapTimeReport = "lap_time_report";
        public const string StatusReport = "status_report";
        public const string RaceStart = "race_start";
        public const string RaceEnd = "race_end";
        public const string LapFeedback = "lap_feedback";
        public const string TrackTrend = "track_trend";
        public const string Fcy = "fcy";
        public const string FcyPitOpen = "fcy_pit_open";
        public const string GreenFlag = "green_flag";
        public const string SectorYellow = "sector_yellow";
        public const string BlueFlag = "blue_flag";
        public const string RaceMilestone = "race_milestone";
        public const string PositionChange = "position_change";
        public const string RivalPit = "rival_pit";
        public const string RivalPace = "rival_pace";
        public const string PitLimiter = "pit_limiter";
        public const string EngineWarning = "engine_warning";
        public const string BrakeWarning = "brake_warning";
        public const string FuelSave = "fuel_save";
    }

    /// <summary>events.py의 Event. C# 예약어/충돌을 피해 RadioEvent.</summary>
    public sealed class RadioEvent
    {
        public string Type;
        public Priority Priority;
        public Dictionary<string, object> Data = new Dictionary<string, object>();
        public string Message;              // 완성된 멘트 텍스트 (있으면 그대로 사용)
        public string DedupKey;             // 없으면 Type이 dedup 키
        public double? Ttl;
        public string Tone = "casual";      // casual | urgent
        public string BridgeTopic;          // 긴급 콜 뒤 LLM 후속 생성 컨텍스트
        public Func<bool> ValidFn;          // 재생 직전 유효성 검사 (false → 폐기)
        public double CreatedAt;            // 버스가 push 시점에 채운다

        public string Key
        {
            get { return DedupKey ?? Type; }
        }

        public bool Expired(double now)
        {
            double ttl = Ttl ?? DefaultTtl(Priority);
            return now - CreatedAt > ttl;
        }

        public static double DefaultTtl(Priority p)
        {
            switch (p)
            {
                case Priority.Critical: return 8.0;   // 긴급 콜은 8초 지나면 의미 없음
                case Priority.High: return 20.0;
                default: return 45.0;
            }
        }
    }

    /// <summary>
    /// 이벤트 버스 (events.py의 EventBus 포팅).
    ///
    /// 시계를 주입받는다 — 실전에선 monotonic 실시간, 리플레이 회귀에선
    /// 스냅샷 기록 시각. 파이썬 하니스도 같은 방식으로 가상 시계를 써서
    /// 두 구현의 쿨다운/TTL 판정이 결정적으로 일치한다.
    /// </summary>
    public sealed class EventBus
    {
        // 타입 → 쿨다운 설정 키 (events.py의 COOLDOWN_KEY)
        private static readonly Dictionary<string, string> CooldownKey =
            new Dictionary<string, string>
            {
                { EventTypes.FuelCritical, "fuel_warning" },
                { EventTypes.FuelWarning, "fuel_warning" },
                { EventTypes.TrafficApproach, "traffic" },
                { EventTypes.TrafficClose, "traffic_close" },
                { EventTypes.TrafficUpdate, "traffic_update" },
                { EventTypes.TrafficMulti, "traffic_multi" },
                { EventTypes.Spotter, "spotter" },
                { EventTypes.BridgeFollowup, "bridge" },
                { EventTypes.PaceComment, "pace_comment" },
                { EventTypes.GapComment, "gap_comment" },
                { EventTypes.TyreWarning, "tyre_warning" },
                { EventTypes.Damage, "damage" },
                { EventTypes.DamageReport, "damage_report" },
                { EventTypes.WheelDamage, "wheel_damage" },
                { EventTypes.Penalty, "penalty" },
                { EventTypes.Fcy, "race_control" },
                { EventTypes.FcyPitOpen, "fcy_pit_open" },
                { EventTypes.GreenFlag, "green_flag" },
                { EventTypes.SectorYellow, "sector_yellow" },
                { EventTypes.BlueFlag, "blue_flag" },
                { EventTypes.RaceMilestone, "race_milestone" },
                { EventTypes.PositionChange, "position_change" },
                { EventTypes.RivalPit, "rival_pit" },
                { EventTypes.RivalPace, "rival_pace" },
                { EventTypes.PitLimiter, "pit_limiter" },
                { EventTypes.EngineWarning, "engine_warning" },
                { EventTypes.BrakeWarning, "engine_warning" },
                { EventTypes.FuelSave, "fuel_save" },
            };

        private static readonly Stopwatch Clock = Stopwatch.StartNew();

        private readonly Dictionary<string, double> _cooldowns;
        private readonly Func<double> _now;
        private readonly object _gate = new object();
        // (priority, seq) 순 정렬 힙 대신 정렬 리스트 — 대기 이벤트는 항상 소수
        private readonly List<RadioEvent> _queue = new List<RadioEvent>();
        private long _seq;
        private readonly Dictionary<RadioEvent, long> _seqOf =
            new Dictionary<RadioEvent, long>();
        private readonly HashSet<string> _pendingKeys = new HashSet<string>();
        private readonly Dictionary<string, double> _lastFired =
            new Dictionary<string, double>();
        private readonly ManualResetEventSlim _available = new ManualResetEventSlim(false);
        private readonly ManualResetEventSlim _urgentPending = new ManualResetEventSlim(false);

        public EventBus(Dictionary<string, double> cooldowns, Func<double> clock = null)
        {
            _cooldowns = cooldowns ?? new Dictionary<string, double>();
            _now = clock ?? (() => Clock.Elapsed.TotalSeconds);
        }

        /// <summary>
        /// push가 수락될 때마다 호출 (락 밖). 리플레이 회귀의 기록 지점이자
        /// 플러그인 UI의 "최근 무전" 소스.
        /// </summary>
        public Action<RadioEvent> Accepted;

        /// <summary>재생 중단 판단용 — CRITICAL 이벤트가 대기 중인가.</summary>
        public bool UrgentPending
        {
            get { return _urgentPending.IsSet; }
        }

        private double CooldownSec(string etype)
        {
            string key;
            if (!CooldownKey.TryGetValue(etype, out key))
                key = etype;
            double sec;
            if (_cooldowns.TryGetValue(key, out sec))
                return sec;
            return _cooldowns.TryGetValue("default", out sec) ? sec : 60.0;
        }

        /// <summary>수락되면 true. 쿨다운/중복으로 버려지면 false.</summary>
        public bool Push(RadioEvent ev)
        {
            double now = _now();
            lock (_gate)
            {
                if (_pendingKeys.Contains(ev.Key))
                    return false;
                string cdKey;
                if (!CooldownKey.TryGetValue(ev.Type, out cdKey))
                    cdKey = ev.Type;
                double last;
                if (_lastFired.TryGetValue(cdKey, out last)
                    && now - last < CooldownSec(ev.Type))
                {
                    // CRITICAL은 같은 쿨다운 그룹의 하위 우선순위보다 한 단계 봐준다:
                    // 쿨다운 절반이 지났으면 통과 (연료 warning 직후 critical 등)
                    if (!(ev.Priority == Priority.Critical
                          && now - last >= CooldownSec(ev.Type) * 0.5))
                    {
                        return false;
                    }
                }
                _lastFired[cdKey] = now;
                _pendingKeys.Add(ev.Key);
                ev.CreatedAt = now;
                _seqOf[ev] = _seq++;
                _queue.Add(ev);
                _available.Set();
                if (ev.Priority == Priority.Critical)
                    _urgentPending.Set();
            }
            Action<RadioEvent> accepted = Accepted;
            if (accepted != null)
            {
                try { accepted(ev); } catch (Exception) { }
            }
            return true;
        }

        /// <summary>우선순위가 가장 높은 유효 이벤트를 꺼낸다. 없으면 null.</summary>
        public RadioEvent Pop(double timeoutSec = 0.5)
        {
            if (!_available.Wait(TimeSpan.FromSeconds(timeoutSec)))
                return null;
            double now = _now();
            lock (_gate)
            {
                while (_queue.Count > 0)
                {
                    RadioEvent ev = TakeBest();
                    _pendingKeys.Remove(ev.Key);
                    if (ev.Expired(now))
                        continue;
                    if (ev.ValidFn != null)
                    {
                        bool valid = true;
                        try
                        {
                            valid = ev.ValidFn();
                        }
                        catch (Exception)
                        {
                            // 유효성 검사 실패는 폐기하지 않는 쪽이 안전
                        }
                        if (!valid)
                            continue;
                    }
                    if (_queue.Count == 0)
                        _available.Reset();
                    if (!AnyCriticalQueued())
                        _urgentPending.Reset();
                    return ev;
                }
                _available.Reset();
                _urgentPending.Reset();
            }
            return null;
        }

        private RadioEvent TakeBest()
        {
            int best = 0;
            for (int i = 1; i < _queue.Count; i++)
            {
                RadioEvent a = _queue[i];
                RadioEvent b = _queue[best];
                if ((int)a.Priority < (int)b.Priority
                    || ((int)a.Priority == (int)b.Priority && _seqOf[a] < _seqOf[b]))
                {
                    best = i;
                }
            }
            RadioEvent ev = _queue[best];
            _queue.RemoveAt(best);
            _seqOf.Remove(ev);
            return ev;
        }

        private bool AnyCriticalQueued()
        {
            for (int i = 0; i < _queue.Count; i++)
            {
                if (_queue[i].Priority == Priority.Critical)
                    return true;
            }
            return false;
        }

        /// <summary>세션 경계에서 호출 — 대기 이벤트와 쿨다운 기록을 모두 비운다.</summary>
        public void Clear()
        {
            lock (_gate)
            {
                _queue.Clear();
                _seqOf.Clear();
                _pendingKeys.Clear();
                _lastFired.Clear();
                _available.Reset();
                _urgentPending.Reset();
            }
        }
    }

    /// <summary>config.py DEFAULTS["cooldowns"] 포팅 + 수다스러움 배율.</summary>
    public static class Cooldowns
    {
        public static Dictionary<string, double> Default(double scale = 1.0)
        {
            var raw = new Dictionary<string, double>
            {
                { "fuel_warning", 240 },
                { "traffic", 20 },
                { "traffic_close", 8 },
                { "traffic_update", 15 },
                { "traffic_multi", 25 },
                { "spotter", 1 },
                { "bridge", 20 },
                { "pace_comment", 120 },
                { "gap_comment", 90 },
                { "tyre_warning", 180 },
                { "damage", 60 },
                { "damage_report", 20 },
                { "wheel_damage", 30 },
                { "penalty", 30 },
                { "lap_analysis", 300 },
                { "stint_briefing", 120 },
                { "lap_feedback", 150 },
                { "track_trend", 600 },
                { "race_control", 15 },
                { "fcy_pit_open", 45 },
                { "green_flag", 30 },
                { "sector_yellow", 45 },
                { "blue_flag", 45 },
                { "race_milestone", 60 },
                { "lap_time_report", 25 },
                { "status_report", 60 },
                { "position_change", 45 },
                { "rival_pit", 90 },
                { "rival_pace", 240 },
                { "pit_limiter", 15 },
                { "engine_warning", 240 },
                { "fuel_save", 180 },
                { "default", 60 },
            };
            if (Math.Abs(scale - 1.0) < 1e-9)
                return raw;
            var scaled = new Dictionary<string, double>(raw.Count);
            foreach (KeyValuePair<string, double> kv in raw)
                scaled[kv.Key] = Math.Round(kv.Value * scale);
            return scaled;
        }
    }
}
